tests/test_hoist.py

```diff
@@ -0,0 +1,264 @@
+    # Analysis
+    scheduler.process(transformation=HoistAnalysis(), reverse=True)
+    # Synthesis
+    scheduler.process(transformation=HoistSynthesis())
+
+    for item in scheduler.items:
+        if write_transformation_output:
+            suffix = '.out.F90'
+            sourcefile = item.source
+            sourcefile.write(path=sourcefile.path.with_suffix(suffix).name)
+        if "driver" in item.name:
+            assert len(item.routine.arguments) == 3
+        elif "kernel1" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "kernel2" in item.name:
+            assert len(item.routine.arguments) == 10
+        elif "device1" in item.name:
+            assert len(item.routine.arguments) == 8
+        elif "device2" in item.name:
+            assert len(item.routine.arguments) == 5
+
+    compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_hoist_find_arrays(here, frontend, config):
+    """
+    Testing hoist functionality with modified *Analysis* part using a `find_variables` method only searching for arrays.
+    """
+
+    class HoistAnalysisFindArrays(HoistAnalysis):
+
+        @staticmethod
+        def find_variables(routine):
+            return [var for var in routine.variables if var not in routine.arguments and isinstance(var, sym.Array)]
+
+    proj = here/'sources/projHoist'
+
+    config['routine'] = [
+        {
+            'name': 'driver',
+            'role': 'driver',
+            'expand': True,
+        },
+    ]
+
+    scheduler = Scheduler(paths=[proj], config=config, seed_routines=['driver'], frontend=frontend)
+
+    # compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+    # Analysis
+    scheduler.process(transformation=HoistAnalysisFindArrays(), reverse=True)
+    # Synthesis
+    scheduler.process(transformation=HoistSynthesis())
+
+    for item in scheduler.items:
+        if write_transformation_output:
+            suffix = '.find_arrays.out.F90'
+            sourcefile = item.source
+            sourcefile.write(path=sourcefile.path.with_suffix(suffix).name)
+        if "driver" in item.name:
+            assert len(item.routine.arguments) == 3
+        elif "kernel1" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "kernel2" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "device1" in item.name:
+            assert len(item.routine.arguments) == 6
+        elif "device2" in item.name:
+            assert len(item.routine.arguments) == 5
+
+    compile_and_test(scheduler=scheduler, here=here, a=(5, 10, 100))
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_hoist_find_variables(here, frontend, config):
+    """
+    Testing hoist functionality with modified *Analysis* part using a `find_variables` method only searching for arrays
+    with variable `a` within the shape/dimensions.
```

Comment:
FYI: the RST markup that Sphinx uses requires double ticks for fixed width formatting, e.g. ` ``a`` `, ` ``find_variables`` `. It's irrelevant for tests but something to be aware of in general.